book-src/ch02-00-let-the-code-begin.md

```diff
@@ -1,3 +1,14 @@
+<!-- 
+   So I know that a lot of programming books start with basic numeric stuff, 
+   but I'm wondering if it might be better to start somewhere else here. 
+   In the introduction, you talk about letters -> words -> sentences -> etc. 
+   Perhaps that might be a good way to start here? Talking about making 
+   strings and substrings or something? Idk, just a thought. Math is the "classic" 
+   way to start, but I always worry in Lisps it'll be off-putting since 
+   "that's not how you write math". 
+
+   It's probably not the best way to go, but I don't know. Just a thought.
+-->
```

Comment:
I was thinking about this as well. I had originally thought to switch what are currently ch2 and 3 - starting with text-based examples and expanding that to basic data structures through the grocery list example.

Right now it starts with math because that was the simplest starting point for Klipse-based examples, in order to introduce the idea of interactive editing and instant feedback early on.

If I did some string manipulation examples to use in Klipse, I'd be open to moving to moving data content to before code content.

What do you think about...
1) basic string manipulation / printing. Would have to introduce operators and lists here for basic function calls
2) saying that strings alone aren't enough, introduce basic data structures through the grocery list
3) use math to introduce user-defined functions and nesting expressions

Reply:
I do like this idea. Although I admit, after sleeping on it I do have second thoughts. I'm really conflicted about the math-based approach vs string based. I think they each have their pros and cons. 
Math-based pros:
- No need to introduce syntax for strings and characters
- Familiar, everyone will be able to understand basic arithmetic
- It flows better into the concept of functions, since basic mathematical functions are learned in school
  - Not a relevant pro if you plan to introduce functions with math after strings anyways
Math-based cons: 
- The syntax is jarring for programmers familiar with c-like languages, and non-programmers alike. 

String based pros: 
- It flows better from the chars -> strings -> sentences -> etc.. analogy used right before
- It can ease people into the syntax before getting hit with the prefix-notation math
- It shows that programming isn't necessarily all about math, which can help keep people interested
String based cons:
- There's more syntax thrown at them right away `\a \b \c` for chars, `"a b c"` for strings
- Need to introduce certain concepts more quickly - lists/vectors

Which one is ultimately the "better" approach, I have *no idea*. Some people will prefer strings, some people will prefer math, and unless it's a "chose your own adventure" and people can do whichever they want first, I think some people are going to be unhappy - which doesn't help at all in choosing what to start with 😂

Reply:
Maybe the analogy at the end of the intro should be changed to start with words instead of letters so the backslash syntax for `\a \b \c` doesn't have to be introduced so early, given that is a bit jarring to Clojure beginners as a whole.

Perhaps, for the sake of flow... Replace the current text-based analogy at the end of the intro with a math-based one in order to transition to ch2. Then, move the text-based analogy to the beginning of ch3 and modify it to explain that in the previous chapter we built up series of computations from small parts, so too can be build up complex data from small parts.